src/hooks/use-snapshots.tsx

```diff
@@ -0,0 +1,376 @@
+
+	useEffect( () => {
+		if ( ! initiated ) {
+			return;
+		}
+		getIpcApi().saveSnapshotsToStorage( snapshots );
+	}, [ snapshots, initiated ] );
+
+	useLayoutEffect( () => {
+		getIpcApi()
+			.getSnapshots()
+			.then( ( storedSnapshots ) => {
+				setSnapshots( storedSnapshots );
+				setInitiated( true );
+			} );
+	}, [] );
+
+	useEffect( () => {
+		if ( ! client?.req || isOffline ) {
+			return;
+		}
+		const deletingSnapshots = snapshots.filter( ( snapshot ) => snapshot.isDeleting );
+		if ( deletingSnapshots.length === 0 ) {
+			setLoadingDeletingAllSnapshots( false );
+			return;
+		}
+		const intervalId = setInterval( async () => {
+			for ( const snapshot of deletingSnapshots ) {
+				if ( snapshot.isDeleting ) {
+					try {
+						const resp: SnapshotStatusResponse = await client.req.get( '/jurassic-ninja/status', {
+							apiNamespace: 'wpcom/v2',
+							site_id: snapshot.atomicSiteId,
+						} );
+						if ( parseInt( resp.is_deleted ) === 1 ) {
+							removeSnapshot( snapshot );
+						}
+					} catch ( error ) {
+						// This error occurs in the background, so we report it but do not
+						// alert the user.
+						Sentry.captureException( error );
+					}
+				}
+			}
+		}, 3000 );
+		return () => {
+			clearInterval( intervalId );
+		};
+	}, [ client?.req, isOffline, removeSnapshot, snapshots ] );
+
+	const addSnapshot = useCallback( ( snapshot: Snapshot ) => {
+		setSnapshots( ( prevSnapshots ) => [ snapshot, ...prevSnapshots ] );
+		setActiveSnapshotCount( ( prevCount ) => prevCount + 1 );
+	}, [] );
+
+	const updateSnapshot = useCallback( ( snapshot: Partial< Snapshot > ) => {
+		setSnapshots( ( snapshots ) => {
+			const index = snapshots.findIndex(
+				( snapshotI ) => snapshotI.atomicSiteId === snapshot.atomicSiteId
+			);
+			if ( index === -1 ) {
+				if ( snapshot.isDeleting ) {
+					const newSnapshots = [ ...snapshots ];
+					newSnapshots.push( snapshot as Snapshot );
+					return newSnapshots;
+				}
+				return snapshots;
+			}
+			const newSnapshots = [ ...snapshots ];
+			newSnapshots[ index ] = { ...snapshots[ index ], ...snapshot };
+			return newSnapshots;
+		} );
+	}, [] );
+
+	const clearFloatingSnapshots = useCallback( function clearFloatingSnapshots(
+		allSnapshots: Pick< Snapshot, 'atomicSiteId' >[]
+	) {
+		const siteIds = allSnapshots.map( ( snapshot ) => snapshot.atomicSiteId );
+		if ( ! siteIds.length ) {
+			setSnapshots( [] );
+			return;
+		}
+		setSnapshots( ( snapshots ) =>
+			snapshots.filter( ( snapshot ) => siteIds.includes( snapshot.atomicSiteId ) )
+		);
+	}, [] );
+
+	useEffect( () => {
+		if ( initiated && ! loadingServerSnapshots && allSnapshots ) {
+			clearFloatingSnapshots( allSnapshots );
+		}
+	}, [ allSnapshots, clearFloatingSnapshots, initiated, loadingServerSnapshots ] );
+
+	// Save snapshots to storage when they change
+	useEffect( () => {
+		if ( ! initiated ) {
+			return;
+		}
+		getIpcApi().saveSnapshotsToStorage( snapshots );
+	}, [ snapshots, initiated ] );
```

Comment:
This code is duplicated in line 99:

https://github.com/Automattic/studio/pull/344/files#diff-494179dc77bcc1f7a2f661db06953be2b984040d47401a3eae9e9f6254385cbcR99-R104

```suggestion
```